Implement PauseVM, ResumeVM, and CreateSnapshot

plamenmpetrov opened the pull request:
Hello, Firecracker 0.23.0 adds new API calls for managing snapshots of a microVM and changing the microVM state. The goal of this PR is to port the PauseVM, ResumeVM, and CreateSnapshot calls to the SDK.

*Issue #, if available:*
#239 

*Description of changes*
- Added relevant PauseVM, ResumeVM, and CreateSnapshot definitions to client/swagger.yaml.
- Generated the updated API client.
- Added wrappers for new API calls in firecracker.go.
- Added new API calls to machine.go

We have not added tests to the PR yet. We are open to feedback on what tests we should add and any other comments in general. We would be glad to finalize and contribute this code.

Authored by @plamenmpetrov and @ustiugov  

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

kzys commented on client/swagger.yaml:
```diff
@@ -398,6 +398,56 @@ paths:
           description: Internal server error
```
How about copying https://github.com/firecracker-microvm/firecracker/blob/v0.23.0/src/api_server/swagger/firecracker.yaml as is? Would it bring other unneeded changes?

plamenmpetrov replied:
I left out the parts relevant to snapshot load and [tracking dirty pages](https://github.com/firecracker-microvm/firecracker/blob/master/src/api_server/swagger/firecracker.yaml#L869). I was not going to include snapshot load support yet as per our [discussion](https://github.com/firecracker-microvm/firecracker-containerd/pull/448) so I decided to leave the relevant API definitions altogether. That being said, copying as-is would not interfere in any way. It is perhaps best to copy as-is so that the swagger file does not need to be updated for version 0.23.0 of the SDK.

kzys replied:
Yeah. It is hard to track what we've copied and what we've left out. Tracking tasks can be done in GitHub issues or somewhere else, rather than this file.

plamenmpetrov replied:
I have copied over the swagger file from firecracker v0.23.0 and updated the API.

xibz commented on machine.go:
```diff
@@ -1002,3 +1002,50 @@ func (m *Machine) setupSignals() {
+	return nil
+}
+
+// CreateSnapshot Creates a snapshot of the VM
```
nit: `CreateSnapshot Creates a snapshot of the VM` -> `CreateSnapshot creates a snapshot of the VM`

plamenmpetrov replied:
Fixed

xibz commented on machine.go:
```diff
@@ -1002,3 +1002,50 @@ func (m *Machine) setupSignals() {
+	return nil
+}
+
+// ResumeVM Resumes the VM
```
nit: `Resumes` -> `resumes`

plamenmpetrov replied:
Fixed

xibz commented on machine.go:
```diff
@@ -1002,3 +1002,50 @@ func (m *Machine) setupSignals() {
 	}()
 }
+
+// PauseVM Pauses the VM
```
nit: `Pauses` -> `pauses`

plamenmpetrov replied:
Fixed

xibz reviewed (changes requested):
Thank you @plamenmpetrov for submitting this change. The changes look good with some minor nit comments. Also, these changes need integration tests. Once those are added, this can be merged.

plamenmpetrov commented:
> Thank you @plamenmpetrov for submitting this change. The changes look good with some minor nit comments. Also, these changes need integration tests. Once those are added, this can be merged.

Hello @xibz , I have added preliminary tests for pause, resume and create snapshot, which seem to pass in the [build](https://buildkite.com/firecracker-microvm/firecracker-go-sdk/builds/1224#17a60bc3-4f45-477b-8f22-5305e3d6619c/495). I also updated the firecracker version used in tests to 0.23.0. However, [TestNetworkMachineCNIWithConfFile](https://buildkite.com/firecracker-microvm/firecracker-go-sdk/builds/1224#17a60bc3-4f45-477b-8f22-5305e3d6619c/531-532) times out. In my understanding, this should not be related to pause, resume, or create snapshot. Perhaps it is caused by another change in firecracker 0.23.0?

xibz reviewed (approved):
Awesome, this looks great! Thank you again for this contribution and updating this to have tests. I reran the tests and they passed. Something upstream must've been broken at the time it was ran
